### ISSUE TITLE
Too much indentation in functions

### ISSUE THREAD

**Issue by sam-s** [emacs-ess/ESS#164]
The recent ESS versions indent like this:
```
a <- lapply(seq(0.5,8,0.3),function (m) {
              ...
            })
```
I prefer the old behavior:
```
a <- lapply(seq(0.5,8,0.3),function (m) {
  ...
})
```
Was the change intentional?
If yes, how do I restore the old behavior?
If not, could you please restore it?
Thanks

**Comment by eddelbuettel** [comment]
Which version were you running?   

If it is the most recent one release **yesterday** could you please upgrade to that version and test?

**Comment by sam-s** [comment]
I pulled from git 10 minutes before reporting the bug

**Comment by eddelbuettel** [comment]
Are you aware of setting such as 

```
;; R-internals manual
;;; ESS
(add-hook 'ess-mode-hook
	  (lambda ()
	    (ess-set-style 'C++ 'quiet)
	    ;; Because
	    ;;                                 DEF GNU BSD K&R C++
	    ;; ess-indent-level                  2   2   8   5   4
	    ;; ess-continued-statement-offset    2   2   8   5   4
	    ;; ess-brace-offset                  0   0  -8  -5  -4
	    ;; ess-arg-function-offset           2   4   0   0   0
	    ;; ess-expression-offset             4   2   8   5   4
	    ;; ess-else-offset                   0   0   0   0   0
	    ;; ess-close-brace-offset            0   0   0   0   0
	    (add-hook 'local-write-file-hooks
		      (lambda ()
			(ess-nuke-trailing-whitespace)))))
```

which I use to copy from the R-Internals manual when I set up a new machine.  Indentation is configurable; you may well be able to configure your desired behaviour.

**Comment by sam-s** [comment]
I know about customization. 
However, the new code broke backwards compatibility.
My question is: was this intentional? 
* if yes, please tell me how to restore the behavior I like
* if no, please fix the bug
thanks!

**Comment by eddelbuettel** [comment]
Maybe you could look at the commit log and start bisecting to where the behaviour changed.  The very recent (bug fix) release addresses one concern, maybe there is a side-effect.  By helping to isolate the change you may make a fix more likely.

**Comment by vspinu** [comment]
> Was the change intentional?

Yes.  The code is now indented with respect to the parent sexp (open paranthesis in your example). This was necessary to fix problems with indentation of consecutive block statements like in this [example](https://github.com/emacs-ess/ESS/blob/master/etc/R-ESS-bugs.R#L666). 

I am not happy with the new ways either.  It wastes too much space. I will see what could be done o revert to old indentation without affecting other stuff.

**Comment by lionel-** [comment]
There is also the issue that `ess-continued-statement-offset` does not seem to be taken into account inside functions. This is what I get when I set this variable to 0:

```r
# outside braces
first %>%
  second %>%
  third

# inside braces
{
  first %>%
    second %>%
      third
}
```

This makes it very difficult to work with pipes or ggplot.

**Comment by lionel-** [comment]
Would it be ok to remove `?\{` from [this condition]
(https://github.com/emacs-ess/ESS/blob/f0c5d8019ead3d8b206ca85f039f6a4841e8ac08/lisp/ess-mode.el#L1267)?

This resolves the bug I mentioned and `ess-brace-imaginary-offset` still works, but this could break other things and I'm not sure what use cases to test.

**Comment by vspinu** [comment]
> but this could break other things and I'm not sure what use cases to test.

It probably breaks something. That condition should be there for a reason there (at least it was at some moment of time) You should check if the change affects `etc/R-ESS-bugs.R` to be reasonably sure about that.

**Comment by lionel-** [comment]
Perfect I'll check out this test file.

I also found a way to correct the issue posted by the OP by adding a setting:
```r
# with ess-indent-block-arg-function set to t
list(arg1 = {
       stuff
     },
     arg2 = {
       stuff
     })

lapply(data, function(x) {
         stuff
       })

# with ess-indent-block-arg-function set to nil
list(arg1 = {
  stuff
}, arg2 = {
  stuff
})

lapply(data, function(x) {
  stuff
})
```

I put these changes in this branch: https://github.com/lionel-/ESS/tree/indent-not-tested

**Comment by lionel-** [comment]
I went over the examples in `etc/R-ESS-bugs.R` and everything looks ok with both modifications.

**Comment by vspinu** [comment]
Coming back to OP's problem. It's rather delicate problem. I would really appreciate people's input.

Consider the following example (indented with current rules): 

```
.a.lst <- list(ex1 = function(p) {
                   cMah <- qchisq(0.975, p)
               },
               ex2 = function(p) {
                   cM <- qchisq(0.95, p)
                   function(d) as.numeric(d < cM)
               })
```

The OP request is to indente after `{` based on the indentation of the previous line  (ESS before 15.03). This makes a lot of sense when there is only one `{` block. But consider the above example:

```

.a.lst <- list(ex1 = function(p) {
    cMah <- qchisq(0.975, p)
},
               ex2 = function(p) {
                   cM <- qchisq(0.95, p)
                   function(d) as.numeric(d < cM)
               })
```

That's rather ugly and was precisely the bug (reported by @mmaechler) that lead to the change in 15.03. 

An alternative would be to shift `ex2` all the way to `0` (this is also what R-studio does):

```
.a.lst <- list(ex1 = function(p) {
    cMah <- qchisq(0.975, p)
},
ex2 = function(p) {
    cM <- qchisq(0.95, p)
    function(d) as.numeric(d < cM)
})
```

**Comment by lionel-** [comment]
Note that the issue only arise if the blocks are on their own lines. e.g., the following has always worked:
```r
list(arg1 = {
  stuff
}, arg2 = {
  more_stuff
}
```

I implemented the alternative suggested by Vitalie in #169. It does make the code more complex.

**Comment by eddelbuettel** [comment]
I think I prefer option 1:

```{.r}
.a.lst <- list(ex1 = function(p) {
                   cMah <- qchisq(0.975, p)
               },
               ex2 = function(p) {
                   cM <- qchisq(0.95, p)
                   function(d) as.numeric(d < cM)
               })
```

**Comment by sam-s** [comment]
My vote goes to Lionel's example as it wastes the least amount of both horizontal and vertical space.
I want to set `ess-indent-block-arg-function` to nil!

**Comment by lionel-** [comment]
I originally made it a setting (`ess-indent-block-arg-function`), but Vitalie felt this should be hard-coded.

**Comment by vspinu** [comment]
> I originally made it a setting (ess-indent-block-arg-function), but Vitalie felt this should be hard-coded.

Yes. If possible would be nice to decide on one default. If not, we just make it configurable. 

The most common usage is with one `{` block like in @sam-s example).  No one noticed the problem (my example `#2`) till very recently. That means it's not very important.  Keeping indentation consistent with R-studio while not high priority,  is  a nice thing to have.  So while option 1 is cleaner I am somewhat tilted towards version `#3`.

Version 1 wastes too much space in the most common case. Now when I write testthat tests I am always annoyed with:

```
test_that("ymd functions correctly parse dates separated by -", {
            expect_that(ymd("2010-01-02"),
                        equals(as.POSIXct("2010-01-02", tz = "UTC")))
            expect_that(ymd("10-01-02"),
                        equals(as.POSIXct("2010-01-02", tz = "UTC")))
            expect_that(ydm("2010-02-01"),
                        equals(as.POSIXct("2010-01-02", tz = "UTC")))
          })
```

Instead of 

```
test_that("ymd functions correctly parse dates separated by -", {
    expect_that(ymd("2010-01-02"),
              equals(as.POSIXct("2010-01-02", tz = "UTC")))
    expect_that(ymd("10-01-02"),
              equals(as.POSIXct("2010-01-02", tz = "UTC")))
    expect_that(ydm("2010-02-01"),
                equals(as.POSIXct("2010-01-02", tz = "UTC")))
})
```

**Comment by sam-s** [comment]
>> No one noticed the problem (my example #2) till very recently.

Nope!!
People are lazy - I know I am.
I have been upset with the new behavior for a long long time, but got my
act together and reported the problem only now.

**Comment by vspinu** [comment]
> I have been upset with the new behavior for a long long time, but got my

I didn't mean your report. I meant the problem reported by @mmaechler back in February 

```
.a.lst <- list(ex1 = function(p) {
    cMah <- qchisq(0.975, p)
},
               ex2 = function(p) {
                   cM <- qchisq(0.95, p)
                   function(d) as.numeric(d < cM)
               })
```

 that eventually lead to the change that you are upset about.

**Comment by mmaechler** [comment]
Thank you Vitalie (@vspinu) .  I've only browsed this too long thread.  I definitely very much loathe the above behavior ("February"), and I think am very close to your and @eddelbuettel 's suggestions.

Honestly, I'm also a bit annoyed by the suggestion that we should follow RStudio "standards".
AFAIK, much of the R parts of the R sources have been written -- or at least reformatted(!) -- with ESS, and we (R Core) have had somewhat basic recommendations about the ESS style to use for R, and at least some R core members had followed that pretty closely. 
If standard, then that would rather be one, and RStudio should follow it.... 
((ok, stopping to whine now))

As Dirk said, this is all a matter of taste, ... and I agree that it would be nice to keep things as customizable as reasonable.

**Comment by lionel-** [comment]
I will make it a setting in the PR then. Vitalie did you mean _not_ indenting? What should be the default exactly?

I also think we need a more systematic way of testing indentation in ESS, so that it's easier to provide changes while making sure we don't break anything. The `R-ESS-bugs.el` file is not satisfactory for this.

I'll send a PR to propose a testing strategy for indentation.

**Comment by vspinu** [comment]
>>> Lionel Henry on Tue, 05 May 2015 09:18:14 -0700 wrote:

 > Vitalie did you mean not indenting

I am moving inline. I have a bunch of comments on your changes.

 > I'll send a PR to propose a testing strategy for indentation.

That would be great! I wanted to do this for ages but never got down to
it.

**Comment by archenemies** [comment]
I got here by Googling - did anyone ever answer the OP's question? I was able to get desirable indentation with `(ess-set-style 'RStudio-)`. It's the only style with `(ess-offset-arguments . prev-line)` - the others have `(ess-offset-arguments . open-delim)`. Actually now I realize that the topic is about indentation inside arguments which are function expressions, but maybe some of these people will enjoy a style which is more space-conservative in general:

    lapply(seq_len(nrow(tab)),
      function(i) tab[i,,drop=F]
    )
